src/OpenTelemetry.Instrumentation.Dependencies/Implementation/GrpcClientDiagnosticListener.cs

```diff
@@ -0,0 +1,94 @@
+﻿// <copyright file="GrpcClientDiagnosticListener.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Instrumentation.Dependencies.Implementation
+{
+    internal class GrpcClientDiagnosticListener : ListenerHandler
+    {
+        private static readonly Regex GrpcMethodRegex = new Regex(@"((?<package>\w+)\.)?(?<service>\w+)/(?<method>\w+)", RegexOptions.Compiled);
+
+        private readonly ActivitySourceAdapter activitySource;
+        private readonly PropertyFetcher startRequestFetcher = new PropertyFetcher("Request");
+
+        public GrpcClientDiagnosticListener(ActivitySourceAdapter activitySource)
+            : base("Grpc.Net.Client", null)
+        {
+            this.activitySource = activitySource;
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+            const string EventNameSuffix = ".OnStartActivity";
+            if (!(this.startRequestFetcher.Fetch(payload) is HttpRequestMessage request))
+            {
+                InstrumentationEventSource.Log.NullPayload(nameof(GrpcClientDiagnosticListener) + EventNameSuffix);
+                return;
+            }
+
+            var grpcMethodTag = activity.Tags.FirstOrDefault(tag => tag.Key == "grpc.method");
+            var grpcMethod = grpcMethodTag.Value?.Trim('/');
+
+            // TODO: Avoid the reflection hack once .NET ships new Activity with Kind settable.
+            activity.GetType().GetProperty("Kind").SetValue(activity, ActivityKind.Client);
```

Comment:
For performance, you should get the property from activity once when the type is created. e.g.

```cs
activityKindPropertyInfo = typeof(Activity).GetProperty("Kind");
```

Then use it here:

```cs
activityKindPropertyInfo.SetValue(activity, ActivityKind.Client);
```

This is faster.

Reply:
👍